DiffEqFlux Blog Post (#274)
JuliaLang/www.julialang.org: blog/_posts/2019-01-18-fluxdiffeq.md

---
layout: post
title:  FluxDiffEq.jl: A Julia Library for Neural Differential Equations
author: Chris Rackauckas, Mike Innes, Yingbo Ma, Lyndon White, Vaibhav Dixit
---

The [Neural Ordinary Differential Equations](https://arxiv.org/abs/1806.07366)

> paper has been a hot topic ever since it has made a splash as Best Paper of

Reviewer: ```suggestion
paper has been a hot topic even before it made a splash as Best Paper of
```

NeuralIPS 2018. By mixing ordinary differential equations and neural networks
they were able to improve training speeds and accuracy over ResNet. Now with the
floodgates opened causing a merge of differential equations merging with machine
learning, the purpose of this blog post is into introduce the reader to
differential equations from a data science perspective and show how to mix these
tools with neural nets.

> The advantages of the Julia DifferentialEquations.jl library for numerically solving

Reviewer: ```suggestion
The advantages of the Julia [DifferentialEquations.jl](https://github.com/JuliaDiffEq/DifferentialEquations.jl) library for numerically solving
```

differential equations have been
[discussed in detail in other posts](http://www.stochasticlifestyle.com/comparison-differential-equation-solver-suites-matlab-r-julia-python-c-fortran/).
Recently, these native Julia differential equation solvers have successfully been embedded

> into the Flux.jl deep learning package, to allow the use of a full suite of

Reviewer: ```suggestion
into the [Flux.jl](https://github.com/FluxML/Flux.jl/) deep learning package, to allow the use of a full suite of
```

highly tested and optimized DiffEq methods within neural networks. Using the new package

> DiffEqFlux.jl, we will show the reader how to easily add differential equation

Reviewer: ```suggestion
[DiffEqFlux.jl](https://github.com/JuliaDiffEq/DiffEqFlux.jl/), we will show the reader how to easily add differential equation
```

Lets link it at least the first time

the amount of bunnies close to constant:

```julia
data = Iterators.repeated((), 100)
opt = ADAM(0.1)
cb = function ()
  display(loss_fd_sde())
  display(plot(solve(remake(prob,p=Flux.data(p)),SOSRI(),saveat=0.1),ylim=(0,6)))
end

# Display the ODE with the current parameter values.
cb()

Flux.train!(loss_fd_sde, params, data, opt, cb = cb)
```

And we can keep going. There are differential equations
[which are piecewise constant](http://docs.juliadiffeq.org/latest/tutorials/discrete_stochastic_example.html)
used in biological simulations, or
[jump diffusion equations from financial models](http://docs.juliadiffeq.org/latest/tutorials/jump_diffusion.html),
and the solvers map right over to the Flux.jl neural network frame work through FluxDiffEq.jl
It is worth noting that FluxDiffEq.jl uses is only around ~50 lines of code to pull this all off.

## Implementing the Neural ODE layer in Julia

Let's go all the way back for a second and now implement the neural ODE layer
in Julia. Remember that this is simply an ODE where the derivative function
is defined by a neural network itself. To do this, let's first define the
neural net for the derivative. We can use a multilayer ...

> ## The core technical challenge: backpropogation through differential equation solvers

Reviewer 2: propogation -> propagation
Twice more below.

with no ODE solver method is this guaranteed to work. For example, here's a quick
equation where a backwards solution to the ODE using the Adams method from the
paper has >1700% error in its final point, even with solver tolerances of 1e-12:

```julia
using Sundials, DiffEqBase
function lorenz(du,u,p,t)
 du[1] = 10.0*(u[2]-u[1])
 du[2] = u[1]*(28.0-u[3]) - u[2]
 du[3] = u[1]*u[2] - (8/3)*u[3]
end
u0 = [1.0;0.0;0.0]
tspan = (0.0,100.0)
prob = ODEProblem(lorenz,u0,tspan)
sol = solve(prob,CVODE_Adams(),reltol=1e-12,abstol=1e-12)
prob2 = ODEProblem(lorenz,sol[end],(100.0,0.0))
sol = solve(prob,CVODE_Adams(),reltol=1e-12,abstol=1e-12)
@show sol[end]-u0 #[-17.5445, -14.7706, 39.7985]
```

(here we once again use the CVODE C++ solvers from SUNDIALS since they are a close
match to the SciPy integrators used in the neural ODE paper)

This inaccuracy is the reason why the method from the neural ODE paper is not
implemented in software suites, but it once again highlights a detail. Not
all ODEs will have a large error due to this issue. And on ODEs where it's not
a problem, this will be the most efficient way to do adjoint sensitivity

> analysis. Thus once again we arive at the conclusion that one method is not

Reviewer 2: arrive

---
layout: post
title:  FluxDiffEq.jl: A Julia Library for Neural Differential Equations
author: Chris Rackauckas, Mike Innes, Yingbo Ma, Lyndon White, Vaibhav Dixit
---

The [Neural Ordinary Differential Equations](https://arxiv.org/abs/1806.07366)
paper has been a hot topic even before it made a splash as Best Paper of
NeuralIPS 2018. By mixing ordinary differential equations and neural networks
they were able to improve training speeds and accuracy over ResNet. Now with the
floodgates opened causing a merge of differential equations merging with machine

> learning, the purpose of this blog post is into introduce the reader to

Reviewer 3: > a merge of differential equations merging

Should merge the "merge" to reduce the "merge"s 😛

---
layout: post
title:  FluxDiffEq.jl: A Julia Library for Neural Differential Equations
author: Chris Rackauckas, Mike Innes, Yingbo Ma, Lyndon White, Vaibhav Dixit
---

In this blog post we
will show you how to easily, efficiently, and
robustly add differential equation solvers to
neural networks in Julia.

![Flux ODE Training Animation](https://user-images.githubusercontent.com/1814174/51399500-1f4dd080-1b14-11e9-8c9d-144f93b6eac2.gif)

The [Neural Ordinary Differential Equations](https://arxiv.org/abs/1806.07366)
paper has been a hot topic even before it made a splash as Best Paper of
NeuralIPS 2018. By mixing ordinary differential equations and neural networks

> they were able to improve training speeds and accuracy over ResNet. Now with the

Reviewer 4: We don't make any claims of improved training speeds or accuracy over a comparably sized ResNet. The win is fewer parameters and O(1) memory to achieve similar performance as ResNet:
![image](https://user-images.githubusercontent.com/4240549/51410744-8169fe00-1b33-11e9-88cc-160eff25a9b4.png)

learning, the purpose of this blog post is into introduce the reader to
differential equations from a data science perspective and show how to mix these
tools with neural nets.

The advantages of the Julia [DifferentialEquations.jl](https://github.com/JuliaDiffEq/DifferentialEquations.jl) library for numerically solving
differential equations have been
[discussed in detail in other posts](http://www.stochasticlifestyle.com/comparison-differential-equation-solver-suites-matlab-r-julia-python-c-fortran/).
Recently, these native Julia differential equation solvers have successfully been embedded
into the [Flux.jl](https://github.com/FluxML/Flux.jl/) deep learning package, to allow the use of a full suite of
highly tested and optimized DiffEq methods within neural networks. Using the new package
[DiffEqFlux.jl](https://github.com/JuliaDiffEq/DiffEqFlux.jl/), we will show the reader how to easily add differential equation
layers to neural networks using a range of differential equations models, including stiff ordinary
differential equations, stochastic differential equations, delay differential
equations, and hybrid (discontinuous) differential equations.

> This is the first user-friendly toolbox to combine state-of-the-art differential

Reviewer 4: Can you qualify this statement with torchdiffeq? Possibly expand 'user-friendly' to distinguish benefits of DiffEq.jl.

Author: Maybe there's a better way of putting this together. I mean for it to kind of go with the previous line, essentially saying this is the first toolbox which combines neural networks with well-optimized and battle tested DE solvers for stiff and non-stiff ODEs, DAEs, DDEs, SDEs, etc. torchdiffeq only does the two basic explicit RK methods (and then a few low order fixed time step methods), so I would think the difference is clear. 

The libraries that would need to be qualified against more IMO are FATODE and Sundials which are also production-quality ODE solvers with adjoint sensitivity analysis implemented, but I think in their case it's clear that they aren't hooked up with neural nets (though they could be).

Author: See if the changed introduction satisfies you. It more qualifies the statement by listing some of the standard features of modern diffeq solver libraries, like distributed parallelism and event handling.

the 5th time point is `0.4` since we set `saveat=0.1`. If we did not set `saveat`,
this would be adaptively chosen. More details on handling the solution type
[can be found in the DifferentialEquations.jl documentation](http://docs.juliadiffeq.org/latest/basics/solution.html).

One last thing to note is that we can make our initial condition (`u0`) and time spans (`tspans`)
to be functions of the parameters (the elements of `p`). For example, we can define the `ODEProblem`:

```julia
u0_f(p,t0) = [p[2],p[4]]
tspan_f(p) = (0.0,10*p[4])
p = [1.5,1.0,3.0,1.0]
prob = ODEProblem(lotka_volterra,u0_f,tspan_f,p)
```

In this form, everything about the problem is determined by the parameter vector (`p`, referred to
as `θ` in associated literature). The utility of this will be seen later.

## Let's Put an ODE Into a Neural Net Framework!

To understand embedding an ODE into a neural network, let's recall what a layer
of a neural network actually is. A layer is a transformation 𝐑ⁿ→𝐑ᵐ which
takes in a vector of size `n` and spits out a new vector of size `m`. That's it!
So a differential equation layer is simply a layer that takes in a parameter

> vector `p` (of size `n`) and spits out a new vector (of size `m`). You can do this

Author: Well `p` can be the input to `u0` and final timespan, that is mentioned later down. So NeuralODE is only using the `p` here for its relation to `u0`, but in a sense it's an expanded form. It does interact oddly with the adjoint method though. @MikeInnes would you recommend having a separate input for these? I believe Flux will differentiate w.r.t. multiple different inputs? The best would be for it to be done via keyword arguments, i.e. `u0 = ..., p = ...`. Then NeuralODE would only use `u0` and other applications would use `p` or both.

> vector `p` (of size `n`) and spits out a new vector (of size `m`). You can do this

Reviewer 4: Sure but even if `p` is `u0` and final timespan it's weird to talk about the dimensionality of the *layer* being that of `p`, imo. Like in this example you wouldn't care about the dimensionality of the timespan.

Is there, in general, some other part of `p` that is changing through time besides `u`? Because when we talk about the layer dimensionality in any other layer we usually mean the dim of the input state, so `u0` and the dim of the output state, `u(t_end)`.  

I know you understand what I'm talking about, just being really explicit here. Is it worth it, in this blog, to be explicit about `p` dimensionality. Or is that a more niche interest? Could we keep it consistent with NeuralODEs in this blog for comparability if that's a reasonable simplifying assumption for an introductory blog?

> vector `p` (of size `n`) and spits out a new vector (of size `m`). You can do this

Author: Is this just a naming thing? What about just calling it `x` instead and noting that `x` is essentially `x = [u0,p,tspan]`, would that make it more clear? Dealing with multiple inputs is probably a better way to handle it if Flux can do that.

>I know you understand what I'm talking about, just being really explicit here. Is it worth it, in this blog, to be explicit about p dimensionality. Or is that a more niche interest?

No, thinking of the solution of an ODE as a function of `p` is definitely not of niche interest. It's the basis of parameter estimation which is by far the biggest use of optimization on ODE solutions. Actually, NeuralODE is very niche in comparison to parameter estimation (which is where the adjoint methods originate from). So for the purposes of integrating data science and ML with differential equations, having the diffeq as both a function of `u0` and `p` are vital to the standard use cases.

> vector `p` (of size `n`) and spits out a new vector (of size `m`). You can do this

Reviewer 5: The confusion here is because Jesse is thinking of optimisation parameters, and Chris is thinking of parameters of the ODE. Probably in all of the neural ODEs work they were the same, but in e.g. the lotka-volterra case they are not; the NN actually generates a vector `p` that gets fed into `f(p)` (which happens to solve an ODE, but isn't a neural ODE in the sense of the paper). In that sense `p` is data in the network rather than an optimisation parameter.

I'm not sure how best to resolve this; one option is to clearly distinguish upfront between the two (and it might be best to describe optimisation parameters as "weights"). I think it's worth spending some time to get this right though, or it'll confuse a lot of people.

> vector `p` (of size `n`) and spits out a new vector (of size `m`). You can do this

Reviewer 6: Definitely good to be explicit upfront about the difference; it confused me too.

let's define a neural network with the function as our single layer, and then a loss
function that is the squared distance of the output values from `1`. In Flux, this looks like:

```julia
p = param([2.2, 1.0, 2.0, 0.4]) # Initial Parameter Vector
params = Flux.Params([p])

function predict_rd() # Our 1-layer neural network
  diffeq_rd(p,reduction,prob,Tsit5(),saveat=0.1)
end

loss_rd() = sum(abs2,x-1 for x in predict_rd()) # loss function
```

Now we tell Flux to train the neural network by running a 100 epoch
to minimise our loss function (`loss_rd()`) and thus obtain the optimized parameters:

```julia
data = Iterators.repeated((), 100)
opt = ADAM(0.1)
cb = function () #callback function to observe training
  display(loss_rd())
  # using `remake` to re-create our `prob` with current parameters `p`
  display(plot(solve(remake(prob,p=Flux.data(p)),Tsit5(),saveat=0.1),ylim=(0,6)))
end

# Display the ODE with the initial parameter values.
cb()

Flux.train!(loss_rd, params, data, opt, cb = cb)
```

The result of this is the animation shown at the top.

Flux.jl finds the parameters of the neural network (`p`) which minimize

> the cost function, i.e. it trains the neural network: it just so happens that here training the neural network happens to include solving an ODE.

Reviewer 4: "training the neural network happens to include solving an ODE" kinda implies that it's the optimizer that requires solving the ODE, not the objective.

amounts of data, differential equations have become an attractive option for
specifying nonlinearities in a learnable (via the parameters) but constrained
form. They are essentially a way of incorporating prior domain-specific knowledge of the structural relations
between the inputs and outputs. Given this way of looking at the two, both methods
tradeoff advantages and disadvantages, making them complementary tools for modeling.
It seems like a clear next step in scientific practice to start putting them
together in new and exciting ways!

## What is the Neural Ordinary Differential Equation (ODE)?

The neural ordinary differential equation is one of many ways to put these two
subjects together. The simplest way of explaining it is that, instead of
learning the nonlinear transformation directly, we wish to learn the structures
of the nonlinear transformation. Thus instead of doing $y=ML(x)$, we put the
machine learning model on the derivative, $y'(x) = ML(x)$, and now solve the ODE.
Why would you ever do this? Well, one motivation is that, if you were to do this
and then solve the ODE using the simplest and most error prone method, the

> Euler method, what you get is equivalent to a [residual neural network](https://arxiv.org/abs/1512.03385).

Author: Maybe it's because ODEs are my bread and butter, but I assumed it's pretty straightforward that if you see it as a method for solving some underlying ODE, you just write out that ODE and solve it another way. The "underlying ODE" part is the convergence of the Euler method as dt->0, which if others think should be mentioned in more detail please submit some text.

> Euler method, what you get is equivalent to a [residual neural network](https://arxiv.org/abs/1512.03385).

Reviewer 5: It's not a question of implementation, but why we should see a model like ResNet as being like solving an ODE in the first place. Most people aren't going to come in with that perspective, or have seen that paper, so it's worth pointing out that we can see ResNet as approximating something "infinitely deep". The idea that deeper models are more powerful gives an intuition for why this gives good results.

The way the Euler method works is based on the fact that $y'(x) = \frac{dy}{dx}$, thus

$$\Delta y = (y_\text{next} - y_\text{prev}) = \Delta x\cdot ML(x) \text{ which implies that } y_{i+1} = y_{i} + \Delta x\cdot ML(x_{i}).$$

Looking familiar now? This is the basis of a residual network (ResNet) which is  type of RNN,
and is the foundation of many of the best AI systems known to date.
The genius of the neural ordinary differential equation paper was to notice this and say,
let's just model the differential equation directly and then solve it using a much better numerical ODE solver methods.
The creation of numerical ODE solvers is a science that goes
all the way back to the first computers, and can adaptively choose step sizes $\Delta x$
and use high order approximations to reduce the number of steps required to get
sufficiently small error. By swapping out the Euler discretization with the LSODE
numerical ODE solver, their results were good enough to make everyone think twice

> about reflexively using an RNN.

Author: Well, saying "make everyone think twice" I think isn't going to far. It's saying that people should start at least considering it. I think it's at least entered the thought cycle.

> about reflexively using an RNN.

Reviewer 5: We should just remove mentions of RNNs unless they are related to the continuous-time recurrence mentioned in the paper. A resnet is not actually a type of RNN anyway, and there's nothing here to show that people should be (or are) throwing out their LSTMs.

> about reflexively using an RNN.

Reviewer: I almost sat down and did the math to show for a suitably general definition of a ResNet, every RNN can be cast as a ResNet.

But @jekbradbury  is 100% right NLP people have not heard of ResNet.

prob2 = ODEProblem(lorenz,sol[end],(100.0,0.0))
sol = solve(prob,CVODE_Adams(),reltol=1e-12,abstol=1e-12)
@show sol[end]-u0 #[-17.5445, -14.7706, 39.7985]
```

(here we once again use the CVODE C++ solvers from SUNDIALS since they are a close
match to the SciPy integrators used in the neural ODE paper)

This inaccuracy is the reason why the method from the neural ODE paper is not
implemented in software suites, but it once again highlights a detail. Not
all ODEs will have a large error due to this issue. And on ODEs where it's not
a problem, this will be the most efficient way to do adjoint sensitivity
analysis. Thus once again we arrive at the conclusion that one method is not
enough.

In DifferentialEquations.jl we are investigating many different methods for
computing the derivatives of differential equations with respect to parameters.
We have a [recent preprint](https://arxiv.org/abs/1812.01892) detailing
some of these results. One of the things we have found is that direct use of
automatic differentiation can be one of the most efficient and flexible methods.
Julia's ForwardDiff.jl, Flux.jl, and ReverseDiff.jl can directly be applied to

> perform automatic differentiation on the native Julia differential equation

Reviewer 6: It's maybe worth mentioning explicitly that this is something only Julia can do well because AD packages in C/C++/Fortran aren't suitable/available/etc. while reimplementing an ODE solver in terms of PyTorch or TensorFlow primitives would be much harder and end up much slower?

Author: Added a paragraph which makes this explicit.

solvers themselves, and this can increase performance while giving new features.
Our findings show that forward-mode automatic differentiation is fastest when
there are less than 100 parameters in the differential equations, and that for

> large enough number of parameters adjoint sensitivity analysis is the most efficient. Even

Reviewer 6: ```suggestion
a large enough number of parameters adjoint sensitivity analysis is the most efficient. Even
```

Please join the [Julia Slack](https://slackinvite.julialang.org/) and the #jsoc channel to discuss in more detail.)

## What do differential equations have to do with machine learning?

The first question someone not familiar with the field might ask is, why are
differential equations important in this context? The simple answer is that a
differential equation is a way to specify an arbitrary nonlinear transform by
mathematically encoding prior structural assumptions.

Let's unpack that statement a bit. There are generally three ways
to define a nonlinear transformation. The first is to explicitly say what it is.
For example, the sigmoid function is $\sigma(x)=\frac{e^x}{e^x + 1}$. This only works
if you know the exact functional form that relates the input to the output.
However, in many cases, such exact relations are not known *a priori*.
So how do you do nonlinear modeling if you don't know the nonlinearity?

There are two separate ways to then handle this problem. One method is machine
learning. Usually this is presented as you have some $x$ and you want to predict
$y$. This generation of a prediction from $x$ is the machine learning model, call it

> $ML$, where you put $x$ into the program and it spits out values $y$. But if you think

Reviewer 7: Suggested wording:
```
There are two common ways to handle this problem. One method is to use machine
learning. In a typical machine learning problem, you are given some input $x$ and 
you want to predict an output $y$. This generation of a prediction $y$ from $x$
is a machine learning model (let's call it $ML$).  During training, we attempt to 
adjust the parameters of $ML$ so that it generates accurate predictions.  We
can then use $ML$ for inference (i.e., produce $y$s for novel inputs $x$).
```

about it differently, this is actually just a nonlinear transformation $y=ML(x)$.

> The reason why the model $ML$ is interesting is because its form is basic but adapts to the

Reviewer 7: `The reason why` => `The reason that`

data itself. For example, a simple neural network (in design matrix form) with
sigmoid activation functions is simply matrix multiplications followed

> by sigmoids, meaning $ML(x)=\sigma(W_{3}\cdot\sigma(W_{2}\cdot\sigma(W_{1}\cdot x)))$ is a three-layer deep

Reviewer 7: `by sigmoids, meaning` => `by application of a sigmoid function.  Specifically, `

opt = ADAM(0.1)
cb = function ()
  display(loss_fd_sde())
  display(plot(solve(remake(prob,p=Flux.data(p)),SOSRI(),saveat=0.1),ylim=(0,6)))
end

# Display the ODE with the current parameter values.
cb()

Flux.train!(loss_fd_sde, params, data, opt, cb = cb)
```

![SDE NN Animation](https://user-images.githubusercontent.com/1814174/51399524-2c6abf80-1b14-11e9-96ae-0192f7debd03.gif)

And we can keep going. There are differential equations
[which are piecewise constant](http://docs.juliadiffeq.org/latest/tutorials/discrete_stochastic_example.html)
used in biological simulations, or
[jump diffusion equations from financial models](http://docs.juliadiffeq.org/latest/tutorials/jump_diffusion.html),
and the solvers map right over to the Flux.jl neural network frame work through FluxDiffEq.jl
It is worth noting that FluxDiffEq.jl uses only around ~50 lines of code to pull this all off.

## Implementing the Neural ODE layer in Julia

Let's go all the way back for a second and now implement the neural ODE layer
in Julia. Remember that this is simply an ODE where the derivative function
is defined by a neural network itself. To do this, let's first define the

> neural net for the derivative. We can use a multilayer ...

Reviewer 7: Is this is meant to be filled in with more details?

Author: yes. That's tomorrow :)

Author: well, today is tomorrow.

  equation solver.
- `diffeq_adjoint` uses adjoint sensitivity analysis to "backprop the ODE solver"

Therefore, to switch from a reverse-mode AD layer to a forward-mode
AD layer, one simply has to change a single character. Since Julia-based automatic
differentiation works on Julia code, the native Julia differential equation
solvers will continue to benefit from advances in this field.

## Conclusion

Machine learning and differential equations are destined to come together due to
their complementary ways of describing a nonlinear world. In the Julia ecosystem
we have merged the differential equation and deep learning packages in such a
way that new developments in the two domains can directly be used together.
We are only beginning to understand the possibilities that have opened up with
this software. We hope that future blog posts will detail some of the cool
applications which mix the two disciplines, such as embedding our coming
pharmacometric simulation engine [PuMaS.jl](https://doi.org/10.1007/s10928-018-9606-9)
into the deep learning framework. With access to the full range of solvers for ODEs,
SDEs, DAEs, DDEs, PDEs, discrete stochastic equations, and more, we are

> interested to see what kinds of next generation neural networks you build with Julia.

Reviewer: ```suggestion
interested to see what kinds of next generation neural networks you will build with Julia.
```

and the solvers map right over to the Flux neural network frame work through FluxDiffEq.jl
FluxDiffEq.jl uses only around ~100 lines of code to pull this all off.

## Implementing the Neural ODE layer in Julia

Let's go all the way back for a second and now implement the neural ODE layer
in Julia. Remember that this is simply an ODE where the derivative function
is defined by a neural network itself. To do this, let's first define the
neural net for the derivative. In Flux, we can define a multilayer perceptron
with 1 hidden layer and a `tanh` activation function like:

```julia
dudt = Chain(Dense(2,50,tanh),Dense(50,2))
```

To define a `neural_ode` layer, we then just need to give
it a timespan and use the `neural_ode` function:

```julia
tspan = (0.0f0,25.0f0)
x->neural_ode(dudt,x,tspan,Tsit5(),saveat=0.1)
```

As a side note, to run this on the GPU, it is sufficient to make the initial
condition and neural network be on the GPU. This will cause the entire ODE
solver's internal operations to take place on the GPU without extra data
transfers in the integration scheme. This looks like[^gpu]:

```julia
x->neural_ode(gpu(dudt),gpu(x),tspan,BS3(),saveat=0.1)
```

> [^gpu]: We have changed the solver from Tsit5 to BS3, as Tsit5 is currently not GPU compatible.

Reviewer 5: That's why I said "currently". Maybe we should add the link over [currently not GPU compatible] so it's more heavily implied.

> [^gpu]: We have changed the solver from Tsit5 to BS3, as Tsit5 is currently not GPU compatible.

Author: This is a delicate subject because then it sounds like it's hard for us to make this GPU compatible. It isn't, we already did it, and we had to revert the PR since v1.0 broadcasting is 3x slower (https://github.com/JuliaLang/julia/issues/28126) and users immediately reported the regression. So we are purposefully choosing to not support GPUs on this algorithm right now because of that, BUT we could add it any time in less than 20 minutes since the code with broadcast is already there and commented out.

How do we politely say this in a way that conveys "okay yes, they essentially have GPU compatibility on everything".

> [^gpu]: We have changed the solver from Tsit5 to BS3, as Tsit5 is currently not GPU compatible.

Reviewer 5: ``
```suggestion
[^gpu]: We have changed the solver from Tsit5 to BS3, as Tsit5 is [currently not GPU compatible]((https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/106)).
```

scatter!(t,Flux.data(pred[1,:]),label="prediction")
```

![Neural ODE Start](https://user-images.githubusercontent.com/1814174/51585822-d9449400-1ea8-11e9-8665-956a16e95207.png)

But now let's train our neural network. To do so, define a prediction function like before, and then
define a loss between our prediction and data:

```julia
function predict_n_ode()
  n_ode(u0)
end
loss_n_ode() = sum(abs2,ode_data .- predict_n_ode())
```

And now we train the neural network and watch as it learns how to
predict our time series:

```julia
data = Iterators.repeated((), 1000)
opt = ADAM(0.1)
cb = function () #callback function to observe training
  display(loss_n_ode())
  # plot current prediction against data
  cur_pred = Flux.data(predict_n_ode())
  pl = scatter(t,ode_data[1,:],label="data")
  scatter!(pl,t,cur_pred[1,:],label="prediction")
  display(plot(pl))
end

# Display the ODE with the initial parameter values.
cb()

Flux.train!(loss_n_ode, ps, data, opt, cb = cb)
```

![Neural ODE Train](https://user-images.githubusercontent.com/1814174/51585825-dc3f8480-1ea8-11e9-8498-18cf55fba3e6.gif)

Notice that what's not being learned is the solution to the ODE.

> Instead, what's learned is the tiny ODE system from which the ODE

Reviewer: ```suggestion
Instead, what we are learning is the tiny ODE system from which the ODE
```

solution is generated. I.e., the neural network inside the neural_ode

> layer has learned this function:

Reviewer: ```suggestion
layer has learns this function:
```

mathematically encoding prior structural assumptions.

Let's unpack that statement a bit. There are generally three ways
to define a nonlinear transformation. The first is to explicitly say what it is.
For example, the sigmoid function is $\sigma(x)=\frac{e^x}{e^x + 1}$. This only works
if you know the exact functional form that relates the input to the output.
However, in many cases, such exact relations are not known *a priori*.
So how do you do nonlinear modeling if you don't know the nonlinearity?

There are two common ways to handle this problem. One method is to use machine
learning. In a typical machine learning problem, you are given some input $x$ and
you want to predict an output $y$. This generation of a prediction $y$ from $x$
is a machine learning model (let's call it $ML$).  During training, we attempt to
adjust the parameters of $ML$ so that it generates accurate predictions.  We
can then use $ML$ for inference (i.e., produce $y$s for novel inputs $x$).
But if you think
about it differently, this is actually just a nonlinear transformation $y=ML(x)$.
The reason $ML$ is interesting is because its form is basic but adapts to the

> data itself. For example, a simple neural network (in design matrix form) with

Reviewer: Design matrix is classical term in statistical regression though.
This is deeplearning folk being badies.

sigmoid activation functions is simply matrix multiplications followed

> by application of sigmoid functions. Specifically, meaning $ML(x)=\sigma(W_{3}\cdot\sigma(W_{2}\cdot\sigma(W_{1}\cdot x)))$ is a three-layer deep

Author: yes

```

![SDE NN Animation](https://user-images.githubusercontent.com/1814174/51399524-2c6abf80-1b14-11e9-96ae-0192f7debd03.gif)

And we can keep going. There are differential equations
[which are piecewise constant](http://docs.juliadiffeq.org/latest/tutorials/discrete_stochastic_example.html)
used in biological simulations, or
[jump diffusion equations from financial models](http://docs.juliadiffeq.org/latest/tutorials/jump_diffusion.html),
and the solvers map right over to the Flux.jl neural network frame work through FluxDiffEq.jl
It is worth noting that FluxDiffEq.jl uses only around ~50 lines of code to pull this all off.

## Implementing the Neural ODE layer in Julia

Let's go all the way back for a second and now implement the neural ODE layer
in Julia. Remember that this is simply an ODE where the derivative function
is defined by a neural network itself. To do this, let's first define the
neural net for the derivative. In Flux, we can define a multilayer perceptron
with 1 hidden layer and a `tanh` activation function like:

```julia
dudt = Chain(Dense(2,50,tanh),Dense(50,2))
```

To define a `neural_ode` layer, we then just need to give it a timespan and use

> the `neural_ode` helper function:

Author: yes, whatever working makes sense to you here. I usually just use "helper function" as "package function you could've easily written yourself in a few lines"

> the `neural_ode` helper function:

Reviewer 6: Yeah, but remember that people coming from PyTorch/TF/NumPy assume that if a function is provided in a package it’s because it would be too difficult to implement correctly themselves.

there are less than 100 parameters in the differential equations, and that for
a large enough number of parameters adjoint sensitivity analysis is the most efficient. Even
then, we have good reason to believe that
[the next generation reverse-mode automatic differentiation via source-to-source AD, Zygote.jl](https://julialang.org/blog/2018/12/ml-language-compiler),
will be more efficient than all of the adjoint sensitivity implementations for
large numbers of parameters.

It is worth noting that using automatic differentiation in this form is almost
impossible for libraries which utilize precompiled shared libraries that call
C/C++/Fortran packages (SciPy, deSolve), and so this is a unique feature of the
Julia differential equation suite. While entire ODE solver libraries could in
theory be recoded into PyTorch or TensorFlow, doing so for a fully featured
suite would be a difficult process without much promise for performance. For
example, it is well-known that production-quality ODE solvers in
C++/Fortran/Julia make heavy use of mutation in first class nonlinear functions
to stave off the heavy memory pressure of allocating small arrays in a loop,
but this kind of workflow is

> [mentioned in the documentation as not well-optimized in PyTorch](https://pytorch.org/docs/master/notes/autograd.html#in-place-operations-with-autograd). Other concerns of this fashion

Reviewer 5: Yeah, I think you can make a much stronger case here by talking about different limitations – e.g. the fact that DE solvers have such a wide range of performance regimes and Python/TF do very badly on scalars.

> [mentioned in the documentation as not well-optimized in PyTorch](https://pytorch.org/docs/master/notes/autograd.html#in-place-operations-with-autograd). Other concerns of this fashion

Author: Mike, you take this then. You know the ML audience's missing knowledge of this domain better.

trade off advantages and disadvantages, making them complementary tools for modeling.
It seems like a clear next step in scientific practice to start putting them
together in new and exciting ways!

## What is the Neural Ordinary Differential Equation (ODE)?

The neural ordinary differential equation is one of many ways to put these two
subjects together. The simplest way of explaining it is that, instead of
learning the nonlinear transformation directly, we wish to learn the structures
of the nonlinear transformation. Thus instead of doing $y=ML(x)$, we put the
machine learning model on the derivative, $y'(x) = ML(x)$, and now solve the ODE.
Why would you ever do this? Well, one motivation is that, if you were to do this
and then solve the ODE using the simplest and most error prone method, the
Euler method, what you get is equivalent to a [residual neural network](https://arxiv.org/abs/1512.03385).
The way the Euler method works is based on the fact that $y'(x) = \frac{dy}{dx}$, thus

$$\Delta y = (y_\text{next} - y_\text{prev}) = \Delta x\cdot ML(x) \text{ which implies that } y_{i+1} = y_{i} + \Delta x\cdot ML(x_{i}).$$

> Looking familiar now? This is the basis of a residual network (ResNet) which is a type of recurrent neural network (RNN),

Reviewer: TO be fair I was working on off some a definition someone else showed me, and they may havebeen wrong or I may have misunderstood.

> Looking familiar now? This is the basis of a residual network (ResNet) which is a type of recurrent neural network (RNN),

Reviewer 2: As far as I know a ResNet is strictly feedforward and claiming that it is in some way recurrent would need a convincing motivation. As written it just risks the credibility of the post.

> Looking familiar now? This is the basis of a residual network (ResNet) which is a type of recurrent neural network (RNN),

Reviewer 6: The paper mentions RNNs but it takes a fairly "DL theory" perspective on what that means. I agree that we should leave it out and focus on the ResNet comparison

> Looking familiar now? This is the basis of a residual network (ResNet) which is a type of recurrent neural network (RNN),

Reviewer 2: If I remember correctly it argued that differential equations could model both ResNet and RNN but not that those two would be otherwise related. I also think focusing on ResNet is a good plan.

and the solvers map right over to the Flux.jl neural network frame work through FluxDiffEq.jl
It is worth noting that FluxDiffEq.jl uses only around ~100 lines of code to pull this all off.

## Implementing the Neural ODE layer in Julia

Let's go all the way back for a second and now implement the neural ODE layer
in Julia. Remember that this is simply an ODE where the derivative function
is defined by a neural network itself. To do this, let's first define the
neural net for the derivative. In Flux, we can define a multilayer perceptron
with 1 hidden layer and a `tanh` activation function like:

```julia
dudt = Chain(Dense(2,50,tanh),Dense(50,2))
```

To define a `neural_ode` layer, we then just need to give it a timespan and use
the `neural_ode` helper function:

```julia
tspan = (0.0f0,25.0f0)
x->neural_ode(dudt,x,tspan,Tsit5(),saveat=0.1)
```

As a side note, to run this on the GPU, it is sufficient to make the initial
condition and neural network be on the GPU. This will cause the entire ODE
solver's internal operations to take place on the GPU without extra data
transfers in the integration scheme. This looks like:

```julia

> x->neural_ode(gpu(dudt),gpu(x),tspan,BS3(),saveat=0.1)

Author: Tsit5 doesn't work on the GPU because no one has fixed v1.0 broadcast regressions yet... :( it literally just needs broadcast added to it but we can't because it would slow it down.

> x->neural_ode(gpu(dudt),gpu(x),tspan,BS3(),saveat=0.1)

Author: `# Currently Tsit5 is purposely not GPU compatible due to Julia v1.0 bug` is that good wording?

applications, but we'll leave those for future discussions.

## The core technical challenge: backpropagation through differential equation solvers

Let's end by explaining the technical issue that needed a solution to make this
all possible. The core to any neural network framework is the ability to
backpropagate derivatives in order to calculate the gradient of the loss function
with respect to the network's parameters. Thus if we stick an ODE solver as a
layer in a neural network, we need to backpropagate through it.

There are multiple ways to do this. The most common is known as (adjoint) sensitivity
analysis. Sensitivity analysis defines a new ODE whose solution gives the
gradients to the cost function w.r.t. the parameters, and solves this secondary
ODE. This is the method discussed in the neural ordinary differential equations
paper, but actually dates back much further, and popular ODE solver frameworks
like [FATODE](http://people.cs.vt.edu/~asandu/Software/FATODE/index.html),
[CASADI](https://web.casadi.org/), and
[CVODES](https://computation.llnl.gov/projects/sundials/cvodes)
have been available with this adjoint method for a long time (CVODES came out

> in 2005!). [DifferentialEquations.jl has sensitivity analysis implemented too](http://docs.juliadiffeq.org/latest/analysis/sensitivity.html)

Reviewer 8: ```suggestion
in 2005!). [DifferentialEquations.jl has sensitivity analysis implemented too](http://docs.juliadiffeq.org/latest/analysis/sensitivity.html)
```
```suggestion
in 2005!) and lastly [DifferentialEquations.jl provides extensive sensitivity analysis implementation](http://docs.juliadiffeq.org/latest/analysis/sensitivity.html)
```

Author: an extensive

a problem, this will be the most efficient way to do adjoint sensitivity
analysis. And this method only applies to ODEs. Not only that, it doesn't even
apply to all ODEs. For example, ODEs with discontinuities ([events](http://docs.juliadiffeq.org/latest/features/callback_functions.html)) are excluded by the assumptions of the derivation.
Thus once again we arrive at the conclusion that one method is not enough.

In DifferentialEquations.jl we have implemented many different methods for
computing the derivatives of differential equations with respect to parameters.
We have a [recent preprint](https://arxiv.org/abs/1812.01892) detailing
some of these results. One of the things we have found is that direct use of
automatic differentiation can be one of the most efficient and flexible methods.
Julia's ForwardDiff.jl, Flux.jl, and ReverseDiff.jl can directly be applied to
perform automatic differentiation on the native Julia differential equation
solvers themselves, and this can increase performance while giving new features.
Our findings show that forward-mode automatic differentiation is fastest when
there are less than 100 parameters in the differential equations, and that for

> a large enough number of parameters adjoint sensitivity analysis is the most efficient. Even

Author: I put 100 in the sentence before.

impossible for libraries which utilize precompiled shared libraries that call
C/C++/Fortran packages (SciPy, deSolve), and so this is a unique feature of the
Julia differential equation suite. While entire ODE solver libraries could in
theory be recoded into PyTorch or TensorFlow, doing so for a fully featured
suite would be a difficult process without much promise for performance. For
example, it is well-known that production-quality ODE solvers in
C++/Fortran/Julia make heavy use of mutation in first class nonlinear functions
to stave off the heavy memory pressure of allocating small arrays in a loop,
but this kind of workflow is
[mentioned in the documentation as not well-optimized in PyTorch](https://pytorch.org/docs/master/notes/autograd.html#in-place-operations-with-autograd). Other concerns of this fashion
make Julia an exceptional choice for this application.

Altogether, being able to switch between different gradient methods without changing
the rest of your code is crucial for having a scalable, optimized, and
maintainable framework for integrating differential equations and neural networks.
And this is precisely what FluxDiffEq.jl gives the user direct access to. There

> are three functions:

Reviewer 7: `There are three functions:` => `There are three functions with a similar API:`

And remove `With a similar API.` below.

![Flux ODE Training Animation](https://user-images.githubusercontent.com/1814174/51399500-1f4dd080-1b14-11e9-8c9d-144f93b6eac2.gif)

The [Neural Ordinary Differential Equations](https://arxiv.org/abs/1806.07366)

> paper, by Jesse Bettencourt and collaborators, has been a hot topic even before it

Reviewer: No, we can't do this.
It is against academic convention to refer to the paper as _by 3rd author and collaborators_ . 

The paper is listed as by
Ricky T. Q. Chen, Yulia Rubanova, Jesse Bettencourt, David Duvenaud

we thus need to say either

by _Ricky T. Q. Chen and collaborators_

or 

_by Ricky T. Q. Chen, Yulia Rubanova, Jesse Bettencourt and David Duvenaud_

Reviewer 5: This is not actually a paper, so it's less important for us to follow that kind of convention unless it's likely to cause frustration for other reasons. I'll try and reword it though.

However, in many cases, such exact relations are not known *a priori*.
So how do you do nonlinear modeling if you don't know the nonlinearity?

> There are two common ways to handle this problem. One method is to use machine

Reviewer: You still need to tell me both of them immediately.
Then you can go into details abotu each.

params = Flux.Params([p])

function predict_rd() # Our 1-layer neural network

>   diffeq_rd(p,prob,Tsit5(),saveat=0.1)

Author: this stuff still needs to do the reduction afterwards or it won't return an array and I'm not sure the code would work.

Reviewer 5: So `diffeq_rd(...)[1,:]`?

Reviewer 5: I'm not totally clear if the slicing does something important here, but conversion to a vector isn't inherently necessary for the AD to work here, so it might still be possible to remove it in some cases.

Author: It's actually going from values in different vectors in a vector of vectors to a single contiguous vector for the time series of 1 variable. So it's not your average slice. It's not important for AD, but if someone puts the layer function in say the `Chain` code and then assumes linear algebra is going to work (putting a Dense right after), they will get an error.
